### PR TITLE
DOPE-202: extended Bucket with Expression and added functionality for…

### PR DESCRIPTION
**Pull request by jansigi**
… aliased Buckets in select and select distinct

### CODE REVIEW

**Review by martinagallati-ergon — APPROVED** [review]

File: core/src/test/kotlin/ch/ergon/dope/resolvable/clause/SelectClauseTest.kt
Comment (pgruntz) [inline core/src/test/kotlin/ch/ergon/dope/resolvable/clause/SelectClauseTest.kt:1]: Please add tests for the case described in the ticket

**Review by pgruntz — CHANGES_REQUESTED** [review]

File: core/src/test/kotlin/ch/ergon/dope/resolvable/clause/SelectClauseTest.kt
Comment (jansigi) [reply core/src/test/kotlin/ch/ergon/dope/resolvable/clause/SelectClauseTest.kt:1]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/SelectClause.kt [hunk 1]
```diff
@@ -50,3 +52,6 @@ class SelectDistinctClause(private val expression: Expression, private vararg va
         )
     }
 }
+
+private fun getExpressionDopeQuery(expression: Expression) =
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/SelectClause.kt:56]: that's looks a bit unclean. You change the generation of the queryString of an AliasedBucket inside the SelectClause. It should be done inside the toDopeQuery-Method of the AliasedBucket.

**Review by pgruntz — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/SelectClause.kt (hunk 1 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/SelectClause.kt:56]: How could I do that?

Comment (pgruntz) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/SelectClause.kt:56]: So the difference is that once it is an _ExpressionBucket_ and once it is a _FromableBucket_ right? So we could split these into separate interfaces/classes for example

Comment (pgruntz) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/SelectClause.kt:56]: Or another approach is making the AliasedBucket not inherit from Expression and add an additional AliasedExpressionBucket

Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/SelectClause.kt]: I added a ReferencedAliasedBucket

File: core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt [hunk 2]
```diff
@@ -22,7 +23,13 @@ class QueryBuilder {
 
     fun selectFrom(fromable: Fromable) = SelectClause(AsteriskExpression()).from(fromable)
 
+    fun selectFrom(bucket: AliasedBucket) = SelectClause(AsteriskExpression()).from(bucket)
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt:26]: name the argument aliasedBucket, otherwise there will be confusion since there exists a type Bucket as well

Comment (pgruntz) [reply core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt:26]: rename it everywhere

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt [hunk 3]
```diff
@@ -2,18 +2,33 @@ package ch.ergon.dope.resolvable.clause.model
...
+class FromClause : ISelectUnnestClause {
+    private val resolvable: Resolvable
+    private val parentClause: ISelectClause
+
+    constructor(fromable: Fromable, parentClause: ISelectClause) {
+        this.resolvable = fromable
+        this.parentClause = parentClause
+    }
+
+    constructor(fromable: AliasedBucket, parentClause: ISelectClause) {
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt:21]: its not a fromable

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt [hunk 4]
```diff
@@ -2,18 +2,33 @@ package ch.ergon.dope.resolvable.clause.model
...
+import ch.ergon.dope.resolvable.Resolvable
 import ch.ergon.dope.resolvable.clause.ISelectClause
 import ch.ergon.dope.resolvable.clause.ISelectUnnestClause
 import ch.ergon.dope.resolvable.formatToQueryStringWithSymbol
+import ch.ergon.dope.resolvable.fromable.AliasedBucket
 import ch.ergon.dope.resolvable.fromable.Fromable
 
-class FromClause(private val fromable: Fromable, private val parentClause: ISelectClause) : ISelectUnnestClause {
+class FromClause : ISelectUnnestClause {
+    private val resolvable: Resolvable
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt:13]: why resolvable? An AliasedBucket is a fromable as well

File: core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt (hunk 2 above)
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt:26]: Why do we need this method anyway? An AliasedBucket is a fromable as well

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt [hunk 5]
```diff
@@ -32,7 +34,7 @@ private enum class OnType {
 
 sealed class SelectJoinClause : ISelectJoinClause {
     private val joinType: JoinType
-    private val bucket: Bucket
+    private val resolvable: Resolvable
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt:37]: Again, why this change?

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt [hunk 6]
```diff
@@ -65,7 +67,40 @@ sealed class SelectJoinClause : ISelectJoinClause {
...
-        this.bucket = bucket
+        this.resolvable = bucket
+        this.onKey = onKey
+        this.forBucket = forBucket
+        this.parentClause = parentClause
+        this.onCondition = null
+        this.onKeys = null
+    }
+
+    constructor(joinType: JoinType, bucket: AliasedBucket, onCondition: TypeExpression<BooleanType>, parentClause: ISelectFromClause) {
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt:78]: I think it's cleaner if you just call the other constructor with changing one argument

**Review by pgruntz — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt (hunk 4 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt:13]: The `ReferencedAliasedBucket` isn't.

File: core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt (hunk 2 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/QueryBuilder.kt:26]: Yes, but we need to know, that it is an AliasedBucket, to add the `AS alias` to it

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt (hunk 3 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt:21]: renamed

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt (hunk 5 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt:37]: same answer

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt (hunk 4 above)
Comment (pgruntz) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt:13]: Can't we move the call ._addBucketReference()_ inside the _toDopeQuery_? then we would get rid of these additional constructors

**Review by pgruntz — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/fromable/Bucket.kt [hunk 7]
```diff
@@ -13,6 +15,15 @@ open class UnaliasedBucket(name: String) : Bucket(name) {
 }
 
 class AliasedBucket(name: String, val alias: String) : Bucket(name) {
+    override fun toDopeQuery(manager: DopeQueryManager) = DopeQuery(
+        queryString = "`$alias`",
+        parameters = emptyMap(),
+    )
+
+    fun addBucketReference() = ReferencedAliasedBucket(name, alias)
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/fromable/Bucket.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/fromable/Bucket.kt [hunk 8]
```diff
@@ -13,6 +15,15 @@ open class UnaliasedBucket(name: String) : Bucket(name) {
...
+}
+
+class ReferencedAliasedBucket(val name: String, val alias: String) : Resolvable {
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/fromable/Bucket.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt (hunk 6 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt (hunk 5 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt (hunk 4 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt]: done

**Review by pgruntz — APPROVED** [review]
